notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,285 @@
+CK machine these days?  I think we originally had it because it was
+pretty simple and easily formalised, and fitted in nicely with the
+reduction semantics.  Maybe that's not so important now?
+
+Another issue is that in the specification we've been skating around
+exactly how builtin evaluation works, but we'll need to be more
+precise (and independent of the implementation language) if and when
+we do away with static builtins.  Maybe the machine implementation
+should be one that's easily specified.  An argument in favour of the
+CK machine was its simplicity, but maybe it's not possible to be very
+simple if we deal with builtins properly.
+
+Also, our primary evaluator will eventually be untyped, and it'll
+presumably be a CEK machine.  We'll surely still want a typed
+evaluator as well though, so maybe that should be the CEK machine as
+well.  It's maybe tempting to use the typed CEK machine and make it a
+bit more robust by using the `TypeArg | TermArg` thing mentioned above
+to count the arities: that'll be slightly less efficient, but we'll
+only need it for the typed evaluator, where efficiency isn't paramount.
+
+
+## Higher-rank builtins
+
+In the specification, builtins are currently allowed to be
+polymorphic, but all the type arguments have to precede the term
+arguments (ie, all the quantifiers appear at the front).  
+
+We've recently had a discussion about whether we want to allow
+type and term arguments to be interleaved, allowing us to
+have builtins with higher-rank types (and also non-canonical
+rank-one types).  The internal machinery doesn't really care
+about this, but it does present some problems.
+
+One possible issue is that if we allow type arguments between term
+arguments there's the question of how we erase them.  If we allow such
+things, then we can have a polymorphic partially-applied builtin, and
+such a thing could be passed to a function which could also accept
+type-abstracted lambdas, and such a function would have to deal
+uniformly with both possibilities.  The difficulty is that we no longer
+have the value restriction, so we have to erase abstractions to
+`delay` and apply `force` when type instantiation occurs.  Since
+there's no possibility of computation occurring when a
+partially-applied builtin is instantiated, it seems strange to
+unnecessarily wrap them in a `delay`. [But what happens if the final
+argument is a type argument, ie, if a builtin returns a polymorphic
+value?  We currently require all builtin arguments to be values and
+builtins are supposed to be parametrically polymorphic _and_ return
+values, so can a builtin really return something that can compute when
+instantiated?  Perhaps it can: maybe we can pass `Λt:*.error` (which
+erases to `delay error`) to the identity function?]
+
+Anyway, perhaps we could modify `force` so that it'll act on _any_
+value, doing nothing when it's not wrapped in `delay`; we would then
```

Comment:
This suggests that we may want kind of untyped type checking: we need to check that `force` is only applied to

1. `delay`ed things
2. builtins where they expect types

For instance, take the following builtin:

```
id : all a. a -> a
```

We erase it to

```
id : delay (any -> any)
```

where

```
delay :: * -> *
(->) :: * -> * -> *
any :: *

any ~= delay any | (->) any any
```

and we also have

```
force : delay any -> any
```

so that

```
force Erased.id 1
```

is "well-typed".

Given that the denotational semantics of untyped lambda calculus is exactly `any ~ (any -> any)`, it might make sense to extend that with isorecursive `delay` & `force`, which should allow us to check that we apply `force` only in appropriate places.

Or maybe we don't want to check anything statically. Anyway, I think we don't need `force` to act on any value, only on a `delay`ed one or on a builtin that before being erased was expecting a type argument at that particular position. Such a check may as well happen at runtime.

Once I start implementing an evaluator for Untyped Plutus Core, I'll implement this exact runtime check.

Reply:
^ That's much sophisticated than I was expecting. Take if then else:
```
ifThenElse : forall a. Bool -> a -> a -> a
```
and an example:
```
[[[{(builtin ifThenElse) (con integer)} (con true)] (con 0)] (con 1)]
```
I thought it would erase to:
```
[[[{force (builtin ifThenElse)} (con true)] (con 0)] (con 1)]
```
i.e. type instantiation gets replaced by `force` and we need to track that there are enough `force`s to match the number of type args, delay doesn't get used.

I will have to think some more about it.

Reply:
> That's much sophisticated than I was expecting. 

It's an alternative idea that is most likely not going to work out. I agree that we should start with what you describe as it makes perfect sense and is easy enough to implement.